File: web/src/resource.ts
import { BitmapFont } from './media/font.js';
import { Animation } from './media/animation.js';

export interface GameImage {
  readonly width: number;
  readonly height: number;
  draw(ctx: CanvasRenderingContext2D, x: number, y: number): void;
  drawScaled(ctx: CanvasRenderingContext2D,
    dx: number, dy: number, dw: number, dh: number,
    sx: number, sy: number, sw: number, sh: number): void;
  subImage(x: number, y: number, w: number, h: number): GameImage;
  getPixel(x: number, y: number): number;
}

class CanvasGameImage implements GameImage {
  public readonly width: number;
  public readonly height: number;
  private readonly canvas: HTMLCanvasElement;
  private readonly ctx: CanvasRenderingContext2D;
  private pixelData: ImageData | null = null;

  constructor(source: HTMLCanvasElement) {
    this.canvas = source;
    this.ctx = source.getContext('2d', { willReadFrequently: true })!;
    this.width = source.width;
    this.height = source.height;
  }

  draw(ctx: CanvasRenderingContext2D, x: number, y: number): void {
    ctx.drawImage(this.canvas, x, y);
  }

  drawScaled(ctx: CanvasRenderingContext2D,
    dx: number, dy: number, dw: number, dh: number,
    sx: number, sy: number, sw: number, sh: number): void {
    ctx.drawImage(this.canvas, sx, sy, sw, sh, dx, dy, dw, dh);
  }

  subImage(x: number, y: number, w: number, h: number): GameImage {
    return new SubImage(this, x, y, w, h);
  }

  getPixel(x: number, y: number): number {
    if (!this.pixelData) {
      this.pixelData = this.ctx.getImageData(0, 0, this.width, this.height);
    }
    const idx = (y * this.width + x) * 4;
    const r = this.pixelData.data[idx]!;
    const g = this.pixelData.data[idx + 1]!;
    const b = this.pixelData.data[idx + 2]!;
    const a = this.pixelData.data[idx + 3]!;
    return (a << 24) | (r << 16) | (g << 8) | b;
  }
}

class SubImage implements GameImage {
  public readonly width: number;
  public readonly height: number;
  private readonly source: GameImage;
  private readonly sx: number;
  private readonly sy: number;
  private cached: CanvasGameImage | null = null;

  constructor(source: GameImage, sx: number, sy: number, sw: number, sh: number) {
    this.source = source;
    this.sx = sx;
    this.sy = sy;
    this.width = sw;
    this.height = sh;
  }

  draw(ctx: CanvasRenderingContext2D, x: number, y: number): void {
    this.source.drawScaled(ctx, x, y, this.width, this.height,
      this.sx, this.sy, this.width, this.height);
  }

  drawScaled(ctx: CanvasRenderingContext2D,
    dx: number, dy: number, dw: number, dh: number,
    sx: number, sy: number, sw: number, sh: number): void {
    this.source.drawScaled(ctx, dx, dy, dw, dh,
      this.sx + sx, this.sy + sy, sw, sh);
  }

  subImage(x: number, y: number, w: number, h: number): GameImage {
    return new SubImage(this.source, this.sx + x, this.sy + y, w, h);
  }

  getPixel(x: number, y: number): number {
    if (!this.cached) {
      const canvas = document.createElement('canvas');
      canvas.width = this.width;
      canvas.height = this.height;
      const ctx = canvas.getContext('2d', { willReadFrequently: true })!;
      this.draw(ctx, 0, 0);
      this.cached = new CanvasGameImage(canvas);
    }
    return this.cached.getPixel(x, y);
  }
}

function mapImagePath(path: string): string {
  return path.replace(/^data\/images\//, 'assets/images/').replace(/\.bmp$/, '.png');
}

function mapTextPath(path: string): string {
  if (path.startsWith('data/rooms/')) return path.replace(/^data\/rooms\//, 'assets/rooms/');
  if (path.startsWith('data/dialogues/')) return path.replace(/^data\/dialogues\//, 'assets/dialogues/');
  return path;
}

// Image processing: tint all visible pixels by color
function processImage(source: HTMLImageElement, tintR: number, tintG: number, tintB: number): CanvasGameImage {
  const canvas = document.createElement('canvas');
  canvas.width = source.naturalWidth;
  canvas.height = source.naturalHeight;
  const ctx = canvas.getContext('2d', { willReadFrequently: true })!;
  ctx.drawImage(source, 0, 0);

  if (tintR !== 255 || tintG !== 255 || tintB !== 255) {
    const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
    const data = imageData.data;
    for (let i = 0; i < data.length; i += 4) {
      if (data[i + 3]! > 0) { // if not transparent
        data[i] = (tintR * data[i]!) / 255;
        data[i + 1] = (tintG * data[i + 1]!) / 255;
        data[i + 2] = (tintB * data[i + 2]!) / 255;
      }
    }
    ctx.putImageData(imageData, 0, 0);
  }

  return new CanvasGameImage(canvas);
}

const preloadedImages = new Map<string, HTMLImageElement>();
const preloadedTexts = new Map<string, string>();
const bitmapCache = new Map<string, GameImage>();
const animationCache = new Map<string, Animation>();
const fontCache = new Map<string, BitmapFont>();
let pendingLoads = 0;
let completedLoads = 0;

export const Resource = {
  preLoad(filename: string): void {
    const url = mapImagePath(filename);
    pendingLoads++;
    const img = new Image();
    img.onload = () => { completedLoads++; };
    img.onerror = () => { console.warn(`Failed to load: ${url}`); completedLoads++; };
    img.src = url;
    preloadedImages.set(filename, img);
  },

  preLoadText(filename: string): void {
    const url = mapTextPath(filename);
    pendingLoads++;
    fetch(url)
      .then(r => r.text())
      .then(text => { preloadedTexts.set(filename, text); completedLoads++; })
      .catch(() => { console.warn(`Failed to load: ${url}`); completedLoads++; });
  },

  preLoadSound(_sound: string): void {
    // Sounds are loaded on-demand by Sound module
    // But count as loaded immediately for preloading purposes
  },

  isDonePreloading(): boolean {
    return completedLoads >= pendingLoads;
  },

  getBitmap(filename: string, tintR = 255, tintG = 255, tintB = 255): GameImage {
    const key = `${filename}:${tintR}:${tintG}:${tintB}`;
    let cached = bitmapCache.get(key);
    if (!cached) {
      const img = preloadedImages.get(filename);
      if (!img || !img.complete) {
        // Return a 1x1 placeholder
        const c = document.createElement('canvas');
        c.width = 1; c.height = 1;
        return new CanvasGameImage(c);
      }
      cached = processImage(img, tintR, tintG, tintB);
      bitmapCache.set(key, cached);
    }
    return cached;
  },

  getText(filename: string): string {
    return preloadedTexts.get(filename) ?? '';
  },

  getAnimation(filename: string, numFrames?: number): Animation {
    const key = `${filename}:${numFrames ?? 'auto'}`;
    let cached = animationCache.get(key);
    if (!cached) {
      cached = new Animation(filename, numFrames);
      animationCache.set(key, cached);
    }
    return cached;
  },

  getFont(filename: string): BitmapFont {
    let cached = fontCache.get(filename);
    if (!cached) {
      cached = new BitmapFont(filename);
      fontCache.set(filename, cached);
    }
    return cached;
  },
};

File: wii/src/resource.cpp
#include "resource.h"
#include "media/animation.h"
#include "media/font.h"
#include <SDL2/SDL_image.h>
#include <fstream>
#include <sstream>
#include <cstdio>

// GameImage implementation
GameImage::GameImage(SDL_Texture* tex, SDL_Surface* surf, int w, int h)
    : texture(tex), width(w), height(h), sx(0), sy(0), sw(w), sh(h), surface(surf) {}

GameImage::GameImage(SDL_Texture* tex, SDL_Surface* surf, int w, int h, int sx_, int sy_, int sw_, int sh_)
    : texture(tex), width(sw_), height(sh_), sx(sx_), sy(sy_), sw(sw_), sh(sh_), surface(surf) {
    (void)w; (void)h;
}

void GameImage::draw(SDL_Renderer* renderer, int x, int y) const {
    if (!texture) return;
    SDL_Rect src = {sx, sy, sw, sh};
    SDL_Rect dst = {x, y, sw, sh};
    SDL_RenderCopy(renderer, texture, &src, &dst);
}

void GameImage::drawScaled(SDL_Renderer* renderer, int dx, int dy, int dw, int dh,
                           int srcX, int srcY, int srcW, int srcH) const {
    if (!texture) return;
    SDL_Rect src = {sx + srcX, sy + srcY, srcW, srcH};
    SDL_Rect dst = {dx, dy, dw, dh};
    SDL_RenderCopy(renderer, texture, &src, &dst);
}

GameImage GameImage::subImage(int x, int y, int w, int h) const {
    return GameImage(texture, surface, width, height, sx + x, sy + y, w, h);
}

uint32_t GameImage::getPixel(int x, int y) const {
    if (!surface) return 0;
    int px = sx + x;
    int py = sy + y;
    if (px < 0 || py < 0 || px >= surface->w || py >= surface->h) return 0;

    SDL_LockSurface(surface);
    uint8_t* pixels = static_cast<uint8_t*>(surface->pixels);
    int bpp = surface->format->BytesPerPixel;
    uint8_t* p = pixels + py * surface->pitch + px * bpp;

    uint32_t pixel = 0;
    switch (bpp) {
        case 1: pixel = *p; break;
        case 2: pixel = *reinterpret_cast<uint16_t*>(p); break;
        case 3:
            if (SDL_BYTEORDER == SDL_BIG_ENDIAN)
                pixel = (p[0] << 16) | (p[1] << 8) | p[2];
            else
                pixel = p[0] | (p[1] << 8) | (p[2] << 16);
            break;
        case 4: pixel = *reinterpret_cast<uint32_t*>(p); break;
    }
    SDL_UnlockSurface(surface);

    uint8_t r, g, b, a;
    SDL_GetRGBA(pixel, surface->format, &r, &g, &b, &a);
    return (static_cast<uint32_t>(a) << 24) | (static_cast<uint32_t>(r) << 16) |
           (static_cast<uint32_t>(g) << 8) | static_cast<uint32_t>(b);
}

// Resource implementation
std::string Resource::mapImagePath(const std::string& path) {
    std::string result = path;
    // "data/images/X.bmp" -> "data/images/X.png"
    auto pos = result.rfind(".bmp");
    if (pos != std::string::npos) {
        result.replace(pos, 4, ".png");
    }
    return result;
}

std::string Resource::mapTextPath(const std::string& path) {
    return path;
}

void Resource::init(SDL_Renderer* renderer) {
    renderer_ = renderer;
    IMG_Init(IMG_INIT_PNG);
}

void Resource::shutdown() {
    for (auto& [key, entry] : images_) {
        if (entry.texture) SDL_DestroyTexture(entry.texture);
        if (entry.surface) SDL_FreeSurface(entry.surface);
    }
    images_.clear();
    bitmapCache_.clear();

    for (auto& [key, anim] : animationCache_) {
        delete anim;
    }
    animationCache_.clear();

    for (auto& [key, font] : fontCache_) {
        delete font;
    }
    fontCache_.clear();

    texts_.clear();
    IMG_Quit();
}

void Resource::preLoad(const std::string& filename) {
    std::string path = mapImagePath(filename);
    if (images_.count(filename)) return;

    SDL_Surface* surface = IMG_Load(path.c_str());
    if (!surface) {
        std::fprintf(stderr, "Failed to load image %s: %s\n", path.c_str(), IMG_GetError());
        images_[filename] = {};
        return;
    }

    SDL_Texture* texture = SDL_CreateTextureFromSurface(renderer_, surface);
    if (!texture) {
        std::fprintf(stderr, "Failed to create texture for %s: %s\n", path.c_str(), SDL_GetError());
        SDL_FreeSurface(surface);
        images_[filename] = {};
        return;
    }

    images_[filename] = {surface, texture};
}

void Resource::preLoadText(const std::string& filename) {
    std::string path = mapTextPath(filename);
    if (texts_.count(filename)) return;

    std::ifstream file(path);
    if (!file.is_open()) {
        std::fprintf(stderr, "Failed to load text %s\n", path.c_str());
        texts_[filename] = "";
        return;
    }

    std::stringstream ss;
    ss << file.rdbuf();
    texts_[filename] = ss.str();
}

bool Resource::isDonePreloading() {
    return true; // Synchronous loading
}

GameImage Resource::getBitmap(const std::string& filename, int tintR, int tintG, int tintB) {
    std::string key = filename + ":" + std::to_string(tintR) + ":" +
                      std::to_string(tintG) + ":" + std::to_string(tintB);
    auto it = bitmapCache_.find(key);
    if (it != bitmapCache_.end()) return it->second;

    auto imgIt = images_.find(filename);
    if (imgIt == images_.end() || !imgIt->second.texture) {
        // Try loading on demand
        preLoad(filename);
        imgIt = images_.find(filename);
        if (imgIt == images_.end() || !imgIt->second.texture) {
            bitmapCache_[key] = {};
            return {};
        }
    }

    auto& entry = imgIt->second;
    int w, h;
    SDL_QueryTexture(entry.texture, nullptr, nullptr, &w, &h);

    if (tintR != 255 || tintG != 255 || tintB != 255) {
        SDL_SetTextureColorMod(entry.texture, tintR, tintG, tintB);
    }

    GameImage img(entry.texture, entry.surface, w, h);
    bitmapCache_[key] = img;
    return img;
}

std::string Resource::getText(const std::string& filename) {
    auto it = texts_.find(filename);
    if (it == texts_.end()) {
        preLoadText(filename);
        it = texts_.find(filename);
        if (it == texts_.end()) return "";
    }
    return it->second;
}

Animation* Resource::getAnimation(const std::string& filename, int numFrames) {
    std::string key = filename + ":" + std::to_string(numFrames);
    auto it = animationCache_.find(key);
    if (it != animationCache_.end()) return it->second;

    auto* anim = new Animation(filename, numFrames);
    animationCache_[key] = anim;
    return anim;
}

BitmapFont* Resource::getFont(const std::string& filename) {
    auto it = fontCache_.find(filename);
    if (it != fontCache_.end()) return it->second;

    auto* font = new BitmapFont(filename);
    fontCache_[filename] = font;
    return font;
}

SDL_Renderer* Resource::getRenderer() {
    return renderer_;
}
